tripy/examples/diffusion/model.py

```diff
@@ -0,0 +1,775 @@
+                x, inputs = block(x, emb)
+            else:
+                x, inputs = block(x, emb, context)
+            saved_inputs += inputs
+
+        x = self.mid_block(x, emb, context)
+
+        for block in self.up_blocks:
+            partial_inputs = saved_inputs[-3:]
+            del saved_inputs[-3:]
+            if isinstance(block, UpBlock2D):
+                x = block(x, emb, partial_inputs)
+            else:
+                x = block(x, emb, context, partial_inputs)
+
+        act = self.conv_out(self.conv_act(self.conv_norm_out(x)))
+        return act
+
+
+class CLIPMLP(tp.Module):
+    def __init__(self):
+        self.fc1 = tp.Linear(768, 3072)
+        self.fc2 = tp.Linear(3072, 768)
+
+    def __call__(self, hidden_states):
+        hidden_states = self.fc1(hidden_states)
+        hidden_states = tp.sigmoid(1.702 * hidden_states) * hidden_states  # quick GELU
+        hidden_states = self.fc2(hidden_states)
+        return hidden_states
+
+
+class CLIPAttention(tp.Module):
+    def __init__(self):
+        self.embed_dim = 768
+        self.num_heads = 12
+        self.head_dim = self.embed_dim // self.num_heads
+        self.k_proj = tp.Linear(self.embed_dim, self.embed_dim)
+        self.v_proj = tp.Linear(self.embed_dim, self.embed_dim)
+        self.q_proj = tp.Linear(self.embed_dim, self.embed_dim)
+        self.out_proj = tp.Linear(self.embed_dim, self.embed_dim)
+
+    def __call__(self, hidden_states, causal_attention_mask):
+        bsz, tgt_len, embed_dim = hidden_states.shape[0], hidden_states.shape[1], hidden_states.shape[2]
+        q, k, v = self.q_proj(hidden_states), self.k_proj(hidden_states), self.v_proj(hidden_states)
+        q, k, v = [
+            tp.transpose(
+                tp.reshape(x, (bsz, tgt_len, self.num_heads, self.head_dim)),
+                1,
+                2,
+            )
+            for x in (q, k, v)
+        ]
+        attn_output = scaled_dot_product_attention(
+            q, k, v, embedding_dim=self.head_dim, attn_mask=causal_attention_mask
+        )
+        out = self.out_proj(tp.reshape(tp.transpose(attn_output, 1, 2), (bsz, tgt_len, embed_dim)))
+        return out
+
+
+class CLIPEncoderLayer(tp.Module):
+    def __init__(self):
+        self.self_attn = CLIPAttention()
+        self.layer_norm1 = tp.LayerNorm(768)
+        self.mlp = CLIPMLP()
+        self.layer_norm2 = tp.LayerNorm(768)
+
+    def __call__(self, hidden_states, causal_attention_mask):
+        residual = hidden_states
+        hidden_states = self.layer_norm1(hidden_states)
+        hidden_states = self.self_attn(hidden_states, causal_attention_mask)
+        hidden_states = residual + hidden_states
+
+        residual = hidden_states
+        hidden_states = self.layer_norm2(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        return hidden_states
+
+
+class CLIPEncoder(tp.Module):
```

Comment:
Could we split up each model into its own file? I think that'll make this a lot easier to read.